# Advanced Certification Program in Computational Data Science
## A program by IISc and TalentSprint
### Mini Project Notebook: Linear Algebra and Calculus

## Problem Statement

 The task is to advise a petroleum company on how to meet the demands of their customers for motor oil, diesel oil and gasoline.

## Learning Objectives

At the end of the experiment, you will be able to

* create arrays and matrices in python
* understand the concepts of linear equations
* solve the system of linear equations

### Data

From a barrel of crude oil, in one day, factory $A$ can produce
* 20 gallons of motor oil,
* 10 gallons of diesel oil, and
* 5 gallons of gasoline

Similarly, factory $B$ can produce
* 4 gallons of motor oil,
* 14 gallons of diesel oil, and
* 5 gallons of gasoline

while factory $C$ can produce
* 4 gallons of motor oil,
* 5 gallons of diesel oil, and
* 12 gallons of gasoline

There is also waste in the form of paraffin, among other things. Factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude, factory $B$ 5 gallons, and factory $C$ 2 gallons.

**Note:** Your conclusion should include a discussion of the nature of the terms *unique*, *no solution*, *overdetermined* and *underdetermined* as they apply in the context of the oil plants.

## Grading = 10 Points

### Create an array

Create an array of size 2x3 with arbitrary values.

In [1]:
import numpy as np

# 2x3 array with arbitrary values
A = np.array([[1, 2, 3], [4, 5, 6]])
print(A)

[[1 2 3]
 [4 5 6]]


### Create the system of Linear Equations

Suppose the current daily demand from distributors is 6600 gallons of motor oil, 5100 gallons of diesel oil and 3100 of gasoline.

Set up the system of equations which describes the above situation. Please include the units as well.

Let the number of barrels used by factory $A$, $B$ and $C$ are $x$, $y$ and $z$ respectively.

Then the system of linear equations will be

$$Motor\ oil:\ \ \ 20x + 4y + 4z = 6600$$

$$Diesel\ oil:\ \ \ 10x + 14y + 5z = 5100$$

$$Gasoline:\ \ \ 5x + 5y + 12z = 3100$$

### Solve the system of Linear Equation (2 points)

How many barrels of crude oil each plant should get in order to meet the demand as a group. Remember that we can only provide each plant with an integral number of barrels.

In [2]:
from sympy import symbols, Eq, solve

x, y, z = symbols('x y z')

eq1 = Eq(20*x + 4*y + 4*z, 6600)
eq2 = Eq(10*x + 14*y + 5*z, 5100)
eq3 = Eq(5*x + 5*y + 12*z, 3100)

solution = solve((eq1, eq2, eq3), (x, y, z))
x = solution[x]
y = solution[y]
z = solution[z]
x= int(x)
y = int(y)
z = int(z)
print(f"Number of barrels required by factory A: {x}")
print(f"Number of barrels required by factory B: {y}")
print(f"Number of barrels required by factory C: {z}")


Number of barrels required by factory A: 287
Number of barrels required by factory B: 128
Number of barrels required by factory C: 85


Suppose the total demand for all products **doubled**. What would the solution now be? How does it compare to the original solution? Why, mathematically, should this have been expected?

In [3]:
from sympy import symbols, Eq, solve

x, y, z = symbols('x y z')

eq1 = Eq(20*x + 4*y + 4*z, 13200)
eq2 = Eq(10*x + 14*y + 5*z, 10200)
eq3 = Eq(5*x + 5*y + 12*z, 6200)

solution = solve((eq1, eq2, eq3), (x, y, z))
x = solution[x]
y = solution[y]
z = solution[z]
x= int(x)
y = int(y)
z = int(z)
print(f"Number of barrels required by factory A: {x}")
print(f"Number of barrels required by factory B: {y}")
print(f"Number of barrels required by factory C: {z}")


Number of barrels required by factory A: 574
Number of barrels required by factory B: 257
Number of barrels required by factory C: 170


Suppose that the company acquires another group of distributors and that the daily demand of this group is 2000 gallons of motor oil, 4000 gallons of gasoline, and 4000 gallons of diesel oil. How would you set up production of just this supply? Are there any options (more than one way)?

In [4]:
from sympy import symbols, Eq, solve

xa, ya, za = symbols('xa ya za')

eq1 = Eq(20*xa + 4*ya + 4*za, 2000)
eq2 = Eq(10*xa + 14*ya + 5*za, 4000)
eq3 = Eq(5*xa + 5*ya + 12*za, 4000)

solution = solve((eq1, eq2, eq3), (xa, ya, za))
xa = solution[xa]
ya = solution[ya]
za = solution[za]
xa= int(xa)
ya = int(ya)
za = int(za)
print(f"Number of barrels of motor oil form acquired company: {xa}")
print(f"Number of barrels of gaoline from aquired company : {ya}")
print(f"Number of barrels of diesel  from aquired company: {za}")

Number of barrels of motor oil form acquired company: 12
Number of barrels of gaoline from aquired company : 187
Number of barrels of diesel  from aquired company: 250


Next, calculate the needs of each factory (in barrels of crude, as usual) to meet the total demand of both groups of distributors. When you have done this, compare your answer to results already obtained. What mathematical conclusion can you draw?

In [5]:
from sympy import symbols, Eq, solve

xt, yt, zt = symbols('xt yt zt')

eq1 = Eq(20*xt + 4*yt + 4*zt, 8600)
eq2 = Eq(10*xt + 14*yt + 5*zt, 9100)
eq3 = Eq(5*xt + 5*yt + 12*zt, 7100)

solution = solve((eq1, eq2, eq3), (xt, yt, zt))
xt = solution[xt]
yt = solution[yt]
zt = solution[zt]
xt= int(xt)
yt = int(yt)
zt = int(zt)
print(f"Number of barrels of motor oil form acquired company: {xt}")
print(f"Number of barrels of gaoline from aquired company : {yt}")
print(f"Number of barrels of diesel  from aquired company: {zt}")

Number of barrels of motor oil form acquired company: 299
Number of barrels of gaoline from aquired company : 316
Number of barrels of diesel  from aquired company: 335


Factory A,B,C---
motor oil: 287
gaoline: 128
diesel: 85
Factory D ---
motor oil: 12
gaoline: 187
diesel: 250

additive property----

287+12 = 299
gaoline 315(round off issue)
disel = 335

### Sensitivity and Robustness (1 point)

In real life applications, constants are rarely ever exactly equal to their stated value; certain amounts of uncertainty are always present. This is part of the reason for the science of statistics. In the above model, the daily productions for the plants would be averages over a period of time. Explore what effect small changes in the parameters have on the output.

To do this, pick any 3 coefficients, one at a time, and increase or decrease them by 3%. For each case , note what effect this has on the solution, as a percentage change. Can you draw any overall conclusion?

lets say Motor oil of factory A is increased by 3%
Motor oil:   20x+4y+4z=6600
so, it will become 20.6x+4y+4z=6600

lets say Diesel oil of factory B is increased by 3%
Diesel oil:   10x+14y+5z=5100
So, it will become 10x+14.42y+5z=5100

lets say Diesel oil of factory C is increased by 3%
Gasoline:   5x+5y+12z=3100
So, it will become 5x+5y+12.36z=3100

In [6]:
from sympy import symbols, Eq, solve

xm, ym, zm = symbols('xm ym zm')

eq1 = Eq(20.6*xm + 4*ym + 4*zm, 6600)
eq2 = Eq(10*xm + 14.42*ym + 5*zm, 5100)
eq3 = Eq(5*xm + 5*ym + 12.36*zm, 3100)

solution = solve((eq1, eq2, eq3), (xm, ym, zm))
xm = solution[xm]
ym = solution[ym]
zm = solution[zm]
xm= int(xm)
ym = int(ym)
zm = int(zm)
print(f"Number of barrels of motor oil form acquired company: {xm}")
print(f"Number of barrels of gaoline from aquired company : {ym}")
print(f"Number of barrels of diesel  from aquired company: {zm}")

Number of barrels of motor oil form acquired company: 278
Number of barrels of gaoline from aquired company : 131
Number of barrels of diesel  from aquired company: 85


Original : motor oil: 287 gaoline: 128 diesel: 85
post increase : motor oil: 278 gaoline: 131 diesel: 85
Factory A (x): −3.23%
Factory B (y): +2.29%
Factory C (z): +0%

### A Plant Off-Line (1 point)

Suppose factory $C$ is shut down by the EPA (Environmental Protection Agency) temporarily for excessive emissions into the atmosphere. If your demand is as it was originally (6600, 5100, 3100), what would you now say about the companies ability to meet it? What do you recommend they schedule for production now?

factory C is shut down
means z= 0

In [7]:
A_reduced = Matrix([
    [20, 4],
    [10, 14],
    [5, 5]
])
b_reduced = Matrix([6600, 5100, 3100])

# Check if overdetermined
A_reduced.col_insert(2, b_reduced)
A_reduced.rref()


NameError: name 'Matrix' is not defined

In [11]:
from sympy import symbols, Eq, solve

x, y = symbols('x y')

eq1 = Eq(20*x + 4*y, 6600)
eq2 = Eq(10*x + 14*y, 5100)
eq3 = Eq(5*x + 5*y, 3100)

solution = solve((eq1, eq2, eq3), (x, y))

# Check if a solution was found
if solution:
    x_val = round(int(solution[x]))  # Access using the symbol 'x'
    y_val = round(int(solution[y]))  # Access using the symbol 'y'
    print(f"Number of barrels required by factory A: {x_val}")
    print(f"Number of barrels required by factory B: {y_val}")
else:
    print("The system of equations has no solution.")


The system of equations has no solution.


### Buying another plant

####(Note the following given information. You will see questions in continuation to this, in the subsequent sections)

This situation has caused enough concern that the CEO is considering buying another plant, identical to the third, and using it permanently. Assuming that all 4 plants are on line, what production do you recommend to meet the current demand (5000, 8500, 10000)? In general, what can you say about any increased flexibility that the 4th plant might provide?

Let the number of barrels used by factory $A$, $B$, $C$ and $D$ are $x$, $y$, $z$ and $w$ respectively.

Then the system of linear equations will be

$$20x + 4y + 4z + 4w = 5000$$

$$10x + 14y + 5z + 5w = 8500$$

$$5x + 5y + 12z + 12w = 10000$$

The above system of linear equation has fewer equations than variables, hence it is *underdetermined* and cannot have a unique solution. In this case, there are either infinitely many solutions or no exact solution. We can solve it by keeping $w$ as constant and using [rref](http://linear.ups.edu/html/section-RREF.html) form to solve the system of linear equation.
Reduced Row Echelon Form

To know about rref implementation in python refer [here](https://docs.sympy.org/latest/tutorial/matrices.html#rref).

In [10]:
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w],
                   [10, 14, 5, 8500-5*w],
                   [5, 5, 12, 10000-12*w]])
# show rref form
A_aug.rref()

(Matrix([
 [1, 0, 0,   195/4],
 [0, 1, 0,  1325/4],
 [0, 0, 1, 675 - w]]),
 (0, 1, 2))

From the above result, it can be seen that 4th plant will share the number of barrels required by the 3rd plant only, while the requirement of 1st and 2nd plant will remain unaffected.

### Calculate the amount of Paraffin supplied (1 point)

The company has just found a candle company that will buy its paraffin. Under the current conditions (i.e, after buying another plant) for demand (5000, 8500, 10000), how much can be supplied to them per day?

According to the problem statement, factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude oil, factory $B$ 5 gallons, and factory $C$ 2 gallons.

Assuming solved values
𝑥,𝑦,𝑧,𝑤
where w = z as w: Factory D (identical to C) so, use:
Paraffin=3x+5y+2z+2w

20x+4y+4z+4w=5000

10x+14y+5z+5w=8500

5x+5y+12z+12w=10000

there are 4 variable but 3 equation , we cannot solve it.
So , we are assuming "w" = "z"

20x+4y+4z+4w=5000 ⇒ 20x+4y+8z=5000

10x+14y+5z+5w=8500 ⇒ 10x+14y+10z=8500

5x+5y+12z+12w=10000 ⇒ 5x+5y+24z=10000
​
  



​





In [12]:
from sympy import symbols, Eq, solve

# Define variables
x, y, z = symbols("x y z")

# Substituting w = z into the system
eq1 = Eq(20*x + 4*y + 8*z, 5000)
eq2 = Eq(10*x + 14*y + 10*z, 8500)
eq3 = Eq(5*x + 5*y + 24*z, 10000)

# Solve the system
solution = solve((eq1, eq2, eq3), (x, y, z))

# Extract solutions
x_val = round(int(solution[x]))
y_val = round(int(solution[y]))
z_val = round(int(solution[z]))
w_val = z_val  # since w = z

# Paraffin calculation
paraffin_A = 3 * x_val
paraffin_B = 5 * y_val
paraffin_C = 2 * z_val
paraffin_D = 2 * w_val
total_paraffin = paraffin_A + paraffin_B + paraffin_C + paraffin_D

# Print results
print(f"x (Factory A): {x_val}")
print(f"y (Factory B): {y_val}")
print(f"z (Factory C): {z_val}")
print(f"w (Factory D): {w_val}")
print(f"Paraffin from A: {paraffin_A}")
print(f"Paraffin from B: {paraffin_B}")
print(f"Paraffin from C: {paraffin_C}")
print(f"Paraffin from D: {paraffin_D}")
print(f"Total Paraffin: {total_paraffin}")

x (Factory A): 48
y (Factory B): 331
z (Factory C): 337
w (Factory D): 337
Paraffin from A: 144
Paraffin from B: 1655
Paraffin from C: 674
Paraffin from D: 674
Total Paraffin: 3147


### Selling the first plant (1 point)

The management is also considering selling the first plant due to aging equipment and high workman's compensation costs for the state it is located in. They would like to know what this would do to their production capability. Specifically, they would like an example of a demand they could not meet with only plants 2 and 3, and also what effect having plant 4 has (recall it is identical to plant 3). They would also like an example of a demand that they could meet with just plants 2 and 3. Any general statements you could make here would be helpful.

Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

When considering only plants 2 and 3, and demand (5000, 8500, 10000) then we have

$$4y + 4z = 5000$$

$$14y + 5z = 8500$$

$$5y + 12z = 10000$$

In [14]:
from sympy import *

y, z = symbols('y z')
eq1 = Eq(4*y + 4*z, 5000)
eq2 = Eq(14*y + 5*z, 8500)
eq3 = Eq(5*y + 12*z, 10000)

sol_without_d = solve([eq1, eq2, eq3], (y, z))
y_val_Aclosed = round(int(solution[y]))
z_val_Aclosed = round(int(solution[z]))
y_val_Aclosed,z_val_Aclosed
if sol_without_d:
    y_val_Aclosed,z_val_Aclosed
else:
    print("The system of equations has no solution.")

The system of equations has no solution.


In [15]:
from sympy import symbols, Eq, solve

# Define symbols
y, z, w = symbols('y z w')

# Equations with plant D
eq1 = Eq(4*y + 4*z + 4*w, 5000)
eq2 = Eq(14*y + 5*z + 5*w, 8500)
eq3 = Eq(5*y + 12*z + 12*w, 10000)

# Solve
sol = solve((eq1, eq2, eq3), (y, z, w))

# Check if sol is a list and handle accordingly
if isinstance(sol, list):
    print("Solution with Plant D (multiple solutions possible):")
    for s in sol:  # Iterate through the list of solutions
        #print(s) # Print each solution as it is since format may vary.
        # If a solution is a dictionary, you can print it similarly as before
        if isinstance(s, dict):
            for var, val in s.items():
                print(f"{var} = {val}")
        else:
            print("Solution format is not a dictionary for this instance.")
            print(f"Solution: {s}")
        print("-----")
else:
    print("Solution with Plant D:")
    for var, val in sol.items():
        print(f"{var} = {val}")

Solution with Plant D (multiple solutions possible):


CONCLUSION: With just B and C, the system may be inconsistent or overconstrained for larger demands.

Adding plant D increases the system’s degrees of freedom, giving more feasible solutions and production resilience.

Taking 4th plant into consideration.
Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

Then for demand (5000, 8500, 10000) the system of linear equations will be

$$4y + 4z + 4w = 5000$$

$$14y + 5z + 5w = 8500$$

$$5y + 12z + 12w = 10000$$

Solve it using rref form.

In [16]:
from sympy import Matrix

# Augmented matrix [A | b]
A_aug = Matrix([
    [4, 4, 4, 5000],
    [14, 5, 5, 8500],
    [5, 12, 12, 10000]
])

# Reduce to RREF
rref_matrix, pivot_columns = A_aug.rref()

# Display
print("RREF Matrix:")
print(rref_matrix)


RREF Matrix:
Matrix([[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 0, 1]])


In [17]:
from sympy import Matrix

# STEP 1: Define the augmented matrix
A_aug = Matrix([
    [4, 4, 4, 5000],
    [14, 5, 5, 8500],
    [5, 12, 12, 10000]
])

# STEP 2: Print shape for debugging
print("Shape of Matrix:", A_aug.shape)  # Should be (3, 4)

# STEP 3: Compute RREF
rref_matrix, pivot_columns = A_aug.rref()

# STEP 4: Display the result
print("RREF Matrix:")
print(rref_matrix)


Shape of Matrix: (3, 4)
RREF Matrix:
Matrix([[1, 0, 0, 0], [0, 1, 1, 0], [0, 0, 0, 1]])


In [ ]:
pivot_columns

Now, changing demand to (6600, 5100, 3100) and solving the system of equation using rref form.

In [18]:
from sympy import Matrix

# Construct the augmented matrix [A | b]
A_aug = Matrix([
    [4, 4, 4, 6600],
    [14, 5, 5, 5100],
    [5, 12, 12, 3100]
])

# Compute RREF
rref_matrix, pivots = A_aug.rref()

# Output result
rref_matrix


Matrix([
[1, 0, 0, 0],
[0, 1, 1, 0],
[0, 0, 0, 1]])

### Set rates for Products (1 point)

Company wants to set the rates of motor oil, diesel oil, and gasoline. For this purpose they have few suggestions given as follows:

* 100, 66, 102 Rupees per gallon,

* 104, 64, 100 Rupees per gallon,

* 102, 68, 98 Rupees per gallon, and

* 96, 68, 100 Rupees per gallon

for motor oil, diesel oil, and gasoline respectively.

Using matrix multiplication, find the rates which result in maximum total price.

Let $M$ denote the matrix such that rows represents different plants (A, B and C), columns represents different products (motor oil, diesel oil and gasoline) and each value represents production of that product from one barrel of crude oil for that plant.

$$M = \begin{bmatrix}
20 & 10 & 5 \\
4 & 14 & 5  \\
 4 & 5 & 12  
\end{bmatrix}$$

Also, $R$ is a matrix having different rates as its columns.

$$R = \begin{bmatrix}
100 & 104 & 102 & 96 \\
66 & 64 & 68 & 68  \\
102 & 100 & 98 & 100  
\end{bmatrix}$$

In [19]:
from sympy import Matrix

# Production matrix M (plants A, B, C x products: motor oil, diesel oil, gasoline)
M = Matrix([
    [20, 10, 5],
    [4, 14, 5],
    [4, 5, 12]
])

# Rate matrix R (motor oil, diesel oil, gasoline in each column configuration)
R = Matrix([
    [100, 104, 102, 96],
    [66, 64, 68, 68],
    [102, 100, 98, 100]
])

In [20]:
# Multiply production and rate matrices
revenue_matrix = M * R

# Display the result
revenue_matrix

Matrix([
[3170, 3220, 3210, 3100],
[1834, 1812, 1850, 1836],
[1954, 1936, 1924, 1924]])

In [22]:
# Sum revenue per rate combination (column-wise)
total_revenues = [sum(revenue_matrix.col(i)) for i in range(revenue_matrix.cols)]

# Get the best one
max_value = max(total_revenues)
best_index = total_revenues.index(max_value) + 1  # +1 for 1-based indexing

In [23]:
max_value

6984

In [24]:
best_index

3

### Marginal Cost (1 point)

The total cost $C(x)$ in Rupees, associated with the production of $x$ gallons of gasoline is given by

$$C(x) = 0.005 x^3 – 0.02 x^2 + 30x + 5000$$

Find the marginal cost when $22$ gallons are produced, where, marginal cost means the instantaneous rate of change of total cost at any level of output.

In [25]:
from sympy import symbols, diff

# Define the symbol and cost function
x = symbols('x')
C = 0.005*x**3 - 0.02*x**2 + 30*x + 5000

# Compute derivative (marginal cost function)
C_prime = diff(C, x)

# Evaluate marginal cost at x = 22
marginal_cost_at_22 = C_prime.subs(x, 22)

print(f"Marginal Cost at x = 22 is: ₹{marginal_cost_at_22:.2f}")

Marginal Cost at x = 22 is: ₹36.38


### Marginal Revenue (1 point)

The total revenue in Rupees received from the sale of $x$ gallons of a motor oil is given by $$R(x) = 3x^2 + 36x + 5.$$

Find the marginal revenue, when $x = 28$, where, marginal revenue means the rate of change of total revenue with respect to the number of items sold at an instant.

In [26]:
from sympy import symbols, diff

# Define the symbol and revenue function
x = symbols('x')
R = 3*x**2 + 36*x + 5

# Compute derivative (marginal revenue)
R_prime = diff(R, x)

# Evaluate marginal revenue at x = 28
marginal_revenue_at_28 = R_prime.subs(x, 28)

print(f"Marginal Revenue at x = 28 is: ₹{marginal_revenue_at_28}")


Marginal Revenue at x = 28 is: ₹204


### Pouring crude oil in tank (1 point)

In a cylindrical tank of radius 10 meter, crude oil is being poured at the rate of 314 cubic meter per hour. Then find

* the rate at which the height of crude oil is increasing in the tank, and
* the height of crude oil in tank after 2 hours.

Formula for volume of a cylinder:
V=πr2h

Rate of change of height
𝑑h
𝑑t
​


In [27]:
import sympy as sp

# Given values
r = 10  # radius in meters
dV_dt = 314  # rate of volume increase in m^3/hour

# Define symbols
t = sp.symbols('t')
h_dt = dV_dt / (sp.pi * r**2)

# Rate of change of height
h_dt_value = h_dt.evalf()
print(f"Rate at which the height is increasing: {h_dt_value} meters per hour")

# Calculate height after 2 hours
h_2_hours = h_dt_value * 2
print(f"Height of crude oil after 2 hours: {h_2_hours} meters")


Rate at which the height is increasing: 0.999493042617103 meters per hour
Height of crude oil after 2 hours: 1.99898608523421 meters
